### PR TITLE
crimson/osd: fix Notify life-time mismanagement in Watch::notify_ack

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: https://tracker.ceph.com/issues/61504

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by rzarzynski** [comment]
We might not need #51936 after this.

**Review by Matan-B — APPROVED** [review]
LGTM, let's verify we no longer hit this issue in the suite's rbd api tests before merging.
I scheduled a build in Shaman.

**Review by athanatos — APPROVED** [review]
This also appears to be fixing a bug where we were calling complete_watcher on all in progress notifies?

**Review by Matan-B — CHANGES_REQUESTED** [review]
`[ RUN      ] TestLibRBD.QuiesceWatchTimeout` is crashing the osds.

```
 0# gsignal in /lib64/libc.so.6
 1# abort in /lib64/libc.so.6
  ..
 4# crimson::osd::Watch::notify_ack(unsigned long, ceph::buffer::v15_2_0::list const&) in ceph-osd
```

See `rbd_api_tests_old_format`/`rbd_api_tests`. I will also try to look at it.
https://pulpito.ceph.com/matan-2023-06-08_06:36:16-crimson-rados-wip-matanb-crimson-testing-bug-61504-distro-crimson-smithi/


Edit: On the positive side, `rbd_python_api_tests_old_format` and `rbd_python_api_tests` look good! It seems that this was responsible for the recent regression with rbd tests.

File: src/crimson/osd/watch.cc
Comment (Matan-B) [inline src/crimson/osd/watch.cc:127]: I suspect this line for causing the failure, perhaps this is not always true that `notify_id` is still in progress?

Comment (Matan-B) [inline src/crimson/osd/watch.cc:127]: This resolves the `TestLibRBD.QuiesceWatchTimeout` for me locally, what do you think?
```suggestion
  if (it == std::end(in_progress_notifies)) {
     logger().debug("notify_ack nothing to do");
     return seastar::now();
  }
```

File: src/crimson/osd/watch.cc
Comment (Matan-B) [inline src/crimson/osd/watch.cc:125]: nit:
```suggestion
  logger().debug("{} gid={} cookie={} notify_id={}",
                 __func__,  get_watcher_gid(), get_cookie(), notify_id);
```

File: src/crimson/osd/watch.cc
Comment (athanatos) [reply src/crimson/osd/watch.cc:127]: Hmm, we shouldn't be getting a notify ack from something not in the in_progress_notifies list, I don't think.  Perhaps make it error() and debug it after this PR merges?

Comment (rzarzynski) [reply src/crimson/osd/watch.cc]: Switched to logging an error.

**Review by Matan-B — APPROVED** [review]
I will open a follow-up tracker to the suppressed issue once this is merged.

**Comment by Matan-B** [comment]
Failures, seem unrelated:

7301383, 7301385, 7301394, 7301396:  https://tracker.ceph.com/issues/61651 (new tracker)
7301384 https://tracker.ceph.com/issues/59165
7301382  `Assertion ptr && *ptr failed`
7301386  `TestInternal.PoolMetadataConfApply` freezes

https://pulpito.ceph.com/matan-2023-06-12_17:29:11-crimson-rados-wip-matanb-crimson-testing-bug-61504-v3-distro-crimson-smithi/

**Comment by Matan-B** [comment]
Follow-up tracker: https://tracker.ceph.com/issues/61652

**Pull request merged**
